qa/rpc-tests/test_framework/mininode.py

```diff
@@ -588,14 +588,22 @@ def to_headers(blocks):
                     lambda: blocks[-1].sha256 in self.getdata_requests, timeout=timeout)
                 assert ok, "did not receive getdata for {}".format(blocks[-1].sha256)
 
-            self.sync_with_ping()
+            if expect_disconnect:
+                self.wait_for_disconnect()
+            else:
+                self.sync_with_ping()
 
```

Comment:
sync with ping doesn't actually work very well in our multi threaded env...not expecting you to change this, just a note to ourselves that we should fix this in another PR at some point.